Controller: database

maksym-bielan opened the pull request:
Operating class parameter should be stored in database along with
channel.

Added operating class in database, also implemented setter and getter
functions.

Signed-off-by: Maksym Bielan <maksym.bielan@globallogic.com>

maksym-bielan committed:
Controller: database. Store operating class in radio node structure

Operating class parameter should be stored in database along with
channel.

Added operating class in database, also implemented setter and getter
functions.

Signed-off-by: Maksym Bielan <maksym.bielan@globallogic.com>

morantr commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -51,7 +51,8 @@ class node {
     std::string manufacturer;
-    int channel = 0;
+    int channel             = 0;
+    uint8_t operating_class = 0;
```
I'm not sure it is the right place to put it. Because it is radio related data, and where you put it is a place for a common parameters (channel, for example, is common to radio and station nodes)

arnout replied:
> I'm not sure it is the right place to put it. Because it is radio related data, and where you put it is a place for a common parameters (channel, for example, is common to radio and station nodes)

It probably isn't the right place to put it, but at the moment it is where the channel is, and it makes sense to put the operating class in the same place I think.

morantr replied:
> > I'm not sure it is the right place to put it. Because it is radio related data, and where you put it is a place for a common parameters (channel, for example, is common to radio and station nodes)
> 
> It probably isn't the right place to put it, but at the moment it is where the channel is, and it makes sense to put the operating class in the same place I think.

I disagree. I don't understand why it doesn't make sense to put it at the right place, under the radio node.

YuriyMasechko replied:
Moved under radio node. Set\get changed accordingly.

morantr commented:
Please use a more descriptive PR headline.

morantr commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -174,6 +174,9 @@ class db {
     int get_node_channel(std::string mac);
 
+    bool set_hostap_operating_class(std::string mac, uint8_t operating_class);
+    int get_hostap_operating_class(std::string mac);
```
If you use a string argument, it probably should be a const reference.
@arnout maybe we should use sMacAdde? What do you think?
Second, I would really like to get rid of "hostap" terminology and use "radio" instead.

arnout replied:
> If you use a string argument, it probably should be a const reference.
> @arnout maybe we should use sMacAdde? What do you think?

Yes please, for any new functionality, use only sMacAddress.

Also, any non-scalar parameter (which std::string definitely is) should use const-ref (cfr. [Coding style, functions, 2nd paragraph](https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#functions)).

> Second, I would really like to get rid of "hostap" terminology and use "radio" instead.

I agree on the principle. However, currently "radio" is used (almost) nowhere, so using it for this would just create more confusion. So I created a new task #436.

YuriyMasechko replied:
Done.

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -232,6 +232,35 @@ int db::get_node_channel(std::string mac)
+    if (!n) {
+        LOG(WARNING) << __FUNCTION__ << " - node " << mac_str << " does not exist!";
+        return 0;
+    } else if (n->get_type() != beerocks::TYPE_SLAVE || n->hostap == nullptr) {
```
```suggestion
    } else if (n->get_type() != beerocks::TYPE_SLAVE || !n->hostap) {
```

smartptrs shouldn't be compared to nullptr, instead operator! should be used.

Same comment below.

YuriyMasechko replied:
Done.

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -232,6 +232,35 @@ int db::get_node_channel(std::string mac)
+        LOG(WARNING) << __FUNCTION__ << " - node " << mac_str << " does not exist!";
+        return 0;
+    } else if (n->get_type() != beerocks::TYPE_SLAVE || n->hostap == nullptr) {
+        LOG(WARNING) << __FUNCTION__ << "node " << mac_str << " is not a valid hostap!";
```
```suggestion
        LOG(WARNING) << "node " << mac_str << " is not a valid hostap!";
```

`LOG()` already prints the function if so configured. The default config doesn't print the function name, but since it prints file and line number it's not needed either.

Same comment below as well.

YuriyMasechko replied:
Set function was removed as not needed.

arnout reviewed (changes requested):
Just adding the functions isn't very helpful. The PR should also have some code that calls those functions.

Since we don't have the controller code yet to get the current operclass and channel, I would just make a stub implementation, i.e. in `db::set_node_channel_bw` set the operating class to 81, 115 or 124 depending on the channel number. With of course a big TODO that the operclass should be an argument of that function.

In fact, I don't think we need `db::set_hostap_operating_class` at all. It is only ever going to be set together with the other channel parameters, so only from `db::set_node_channel_bw` or its replacement. So, for this first commit I would move the setting to `db::set_node_channel_bw`. In a second commit, the setter should be used for the steering request. If there's any other place where we need the current operating class (which I don't think is the case at the moment), subsequent commits should use operclass there as well.

YuriyMasechko committed:
controller: set dummy operating class based on channel

Set dummy operating class based on channel. Later operating class
should be function parameter.

Set operating class based on channel.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

YuriyMasechko committed:
controller: get operatig class from db instead of hardcode

Steering task should get operating class from db and not
from hardcode.

Get operating class from db.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>
